File: cpp_code/graph/bfs/CC.cpp
/**
* 抖码算法，让算法学习变的简单有趣
* @作者 : 老汤
*/

#include <queue>
#include "../AdjSet.cpp"

class CC {
public:
    CC(AdjSet g): g(g) {
        this->visited.resize(g.getV());
        std::fill(this->visited.begin(), this->visited.end(), -1);

        // 遍历图中每个顶点
        for (int v = 0; v < g.getV(); ++v) {
            // 先判断，没有遍历的顶点才能进行深度优先遍历
            if (visited[v] == -1) {
                ccCount++;
                bfs(v, ccCount);
            }
        }
    }

    int getCcCount() {
        return ccCount;
    }

    vector<vector<int>> components() {
        vector<vector<int>> res;
        for (int i = 0; i < ccCount; ++i) {
            vector<int> temp;
            res.push_back(temp);
        }
        for (int v = 0; v < g.getV(); ++v) {
            int cc = visited[v];
            res[cc - 1].push_back(v);
        }
        return res;
    }

    bool isConnected(int v, int w) {
        validateVertex(v);
        validateVertex(w);
        return visited[v] == visited[w];
    }

private:
    AdjSet g;
    int ccCount = 0;
    // 用于防止一个节点被重复访问
    vector<int> visited;

    void bfs(int v, int ccId) {
        queue<int> q;
        q.push(v);
        // 维护节点所属的连通分量
        visited[v] = ccId;

        while (!q.empty()) {
            int curr = q.front();
            q.pop();
            for (auto& w : g.adj(curr)) {
                if (visited[w] == -1) {
                    q.push(w);
                    // 维护节点所属的连通分量
                    visited[w] = ccId;
                }
            }
        }
    }

    void validateVertex(int v) {
        if (v < 0 || v >= g.getV()) {
            throw "顶点 " + to_string(v) + " 不合格";
        }
    }
};

int main() {
    AdjSet g("C:/workspace/douma-algo/data/graph-bfs.txt");
    CC cc(g);
    cout << to_string(cc.getCcCount()) << endl;
    return 0;
}

File: javascript_code/graph/bfs/CC.js


export class CC {

    constructor(g) {
        this.g = g
        // 除了标记是否被访问过这个信息外，还记录每个节点所属的连通分量
        this.visited = new Array(g.getV()).fill(-1)
        this.ccCount = 0

        for (let v = 0; v < g.getV(); v++) {
            if (this.visited[v] == -1) {
                this.ccCount++
                this.bfs(v, this.ccCount)
            }
        }
    }

    bfs(v, ccId) {
        const queue = []
        queue.push(v)
        // 维护节点所属的连通分量
        this.visited[v] = ccId
        while (queue.length) {
            const curr = queue.shift()

            for (const w of this.g.adj(curr)) {
                if (this.visited[w] == -1) {
                    queue.push(w)
                    // 维护节点所属的连通分量
                    this.visited[w] = ccId
                }
            }
        }
    }

    getCCCount() {
        return this.ccCount
    }

    components() {
        const res = new Array(this.ccCount)
        for (let i = 0; i < this.ccCount; i++) {
            res[i] = []
        }
        for (let v = 0; v < this.g.getV(); v++) {
            const cc = this.visited[v]
            res[cc - 1].push(v)
        }
        return res
    }

    isConnected(v, w) {
        this.validateVertex(v)
        this.validateVertex(w)
        return this.visited[v] == this.visited[w]
    }

    validateVertex(v) {
        if (v < 0 && v >= this.g.getV()) {
            throw new Error("顶点不合法，超出范围")
        }
    }
}